kotlinx-coroutines-io/src/main/kotlin/kotlinx/coroutines/experimental/io/Coroutines.kt

```diff
@@ -128,9 +106,32 @@ fun writer(
     }
 }
 
-internal class ReaderCoroutine(context: CoroutineContext, channel: ByteChannel) :
-    ByteChannelCoroutine(context, channel), ReaderJob, ReaderScope
+/**
+ * @param S not exactly safe (unchecked cast is used) so should be [ReaderScope] or [WriterScope]
+ */
+private fun <S : CoroutineScope> CoroutineScope.launchChannel(context: CoroutineContext,
+                                                              channel: ByteChannel,
+                                                              attachJob: Boolean,
+                                                              block: suspend S.() -> Unit): ChannelJob {
+    val job = launch(context) {
+        if (attachJob) {
+            channel.attachJob(coroutineContext[Job]!!)
+        }
+        try {
+            @Suppress("UNCHECKED_CAST")
+            block(ChannelScope(this, channel) as S)
+        } catch (t: Throwable) {
+            channel.cancel(t)
+            throw CancellationException("Cancelled due to error $t")
+        } finally {
+            channel.close()
```

Comment:
If `block` spawns children, then the channel can be closed before they complete.
Is it intended behaviour?

Reply:
Same about `cancel` call

Reply:
You are right, perhaps need to do it in regular `invokeOnCompletion`, thanks